src/seraphis_wallet/serialization_types.h

```diff
@@ -0,0 +1,129 @@
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+// local headers
+#include "ringct/rctTypes.h"
+#include "seraphis_core/jamtis_destination.h"
+#include "seraphis_crypto/sp_crypto_utils.h"
+#include "seraphis_impl/enote_store.h"
+#include "seraphis_impl/serialization_demo_types.h"
+#include "seraphis_main/contextual_enote_record_types.h"
+#include "seraphis_main/sp_knowledge_proof_types.h"
+#include "seraphis_main/sp_knowledge_proof_utils.h"
+
+// third party headers
+#include <boost/range.hpp>
+
+#include "boost/range/iterator_range.hpp"
+#include "seraphis_wallet/transaction_history.h"
+#include "serialization/binary_archive.h"
+#include "serialization/containers.h"
+#include "serialization/serialization.h"
+
+// standard headers
+#include <cstddef>
+#include <cstdint>
+#include <functional>
+#include <map>
+#include <tuple>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+using namespace sp::knowledge_proofs;
+using namespace sp::serialization;
+
+struct ser_TransactionRecordV1
+{
+    // key images of spent enotes for tracking purposes
+    std::vector<crypto::key_image> legacy_spent_enotes;
+    std::vector<crypto::key_image> sp_spent_enotes;
+
+    // sent funds
+    std::vector<std::pair<sp::serialization::ser_JamtisDestinationV1, rct::xmr_amount>> outlays;
+
+    // fees and total sent:
+    rct::xmr_amount amount_sent;
+    rct::xmr_amount fee_sent;
+
+    BEGIN_SERIALIZE_OBJECT()
+    FIELD(legacy_spent_enotes)
+    FIELD(sp_spent_enotes)
+    FIELD(outlays)
+    FIELD(amount_sent)
+    FIELD(fee_sent)
+    END_SERIALIZE()
+};
+
+struct ser_SpTransactionStoreV1
+{
+    // quickly find TransactionRecordV1 from txid
+    serializable_unordered_map<rct::key, ser_TransactionRecordV1> tx_records;
+
+    // sort by blockheight to find last transactions or txs
+    // in a specific time range
+    serializable_multimap<std::uint64_t, rct::key, std::greater<std::uint64_t>> confirmed_txids;
+
+    // sort by timestamp instead of blockheight
+    serializable_multimap<std::uint64_t, rct::key, std::greater<std::uint64_t>> unconfirmed_txids;
+    serializable_multimap<std::uint64_t, rct::key, std::greater<std::uint64_t>> offchain_txids;
+
+    BEGIN_SERIALIZE_OBJECT()
+    FIELD(tx_records)
```

Comment:
Usually those `FIELD` lines get indented 1 level, and that seems only logical to me: Technically `BEGIN_SERIALIZE_OBJECT()` expands to a method, and those `FIELD`s to statements within that method.